Add VMware documentation (#1927)
giantswarm/docs: src/content/advanced/vsphere-kubevip/index.md

---
linkTitle: vSphere Kube-vip
title: vSphere networking with Kube-vip
description: Here we describe how to work with Kube-vip in Giant Swarm CAPV clusters.
weight: 25
menu:
  main:
    parent: advanced
last_review_date: 2023-10-09
user_questions:
  - How to slice my subnet for Kube-vip with CAPV?
  - What IP should I use for Kube-vip with CAPV?
  - How to set a range of IPs for load balancer in Kube-vip with CAPV?
  - Where can I see the IPs in use by Kube-vip?
aliases:
  - /guides/kube-vip-for-capv/
owner:
  - https://github.com/orgs/giantswarm/teams/team-rocket
---

> While most infrastructure providers offer a load balancer to use for the Kubernetes API and services of type Load Balancer (AWS, Azure, VMware Cloud Director...), vSphere is a bit different in that it is a hypervisor as opposed to a cloud provider. Advanced networking use cases such as routing, firewalling, load balancing and the likes are handled by VMware NSX. In order to offer multi-control plane nodes to all vSphere customers (even those without NSX), CAPV comes with `Kube-vip`, a layer-2 load balancer.

Reviewer: > In order to offer multi-control plane nodes to all vSphere customers

Not sure if "multi-control plane nodes" is clear for customers. Maybe:

> In order to offer a highly-available Kubernetes API even to customers without NSX, CAPV comes with kube-vip, a layer-2 load balancer.

Author: I agree

> As a result of working on layer 2, `Kube-vip` uses IP addresses in the same subnet as the nodes themselves and these IP addresses must be selected carefully to avoid collisions. This is particularly plausible if multiple operators collaborate on the platform.

Reviewer: > This is particularly plausible if multiple operators collaborate on the platform.

Which operators? Humans, software? Let's rather give an example so customers can understand in which cases there's really a risk.

Author: done

## IPAM controller

> in order to mitigate this risk, Giant Swarm CAPV clusters run `cluster-api-ipam-provider-in-cluster` to track IP addresses usage from within the management cluster (MC). For that reason, **we recommend running a maximum of one management cluster per subnet**.

Reviewer: ```suggestion
in order to mitigate the risk of IP collisions, Giant Swarm CAPV clusters run `cluster-api-ipam-provider-in-cluster` to track IP addresses usage from within the management cluster (MC). For that reason, **we recommend running a maximum of one management cluster per subnet**.
```

(so we don't expect people to have read the above paragraph)

And customers don't really run the management clusters, right? Shouldn't the warning say that they shouldn't _use_ different managaement clusters to provision workload clusters in the same subnet?

Author: 👍

The basic operations of `cluster-api-ipam-provider-in-cluster` work as follows:

* Create a pool of IPs with a `globalinclusterippools` custom resource (CR).
* Claim an IP by referencing the pool in a `ipaddressclaims` CR.
* An `ipaddresses` CR is created and the `Free/Used` status of the `globalinclusterippools` CR is updated.

```sh

> > k get globalinclusterippools.ipam.cluster.x-k8s.io

Reviewer: Please replace `k` with `kubectl` everywhere, `>` with `$`, and since such code blocks contain output, they should be marked with `nohighlight` formatting (not `sh`)

NAME        ADDRESSES                         TOTAL   FREE   USED
wc-cp-ips   ["10.10.222.232-10.10.222.239"]   8       7      1

> k get ipaddressclaims.ipam.cluster.x-k8s.io -A
NAMESPACE        NAME    POOL NAME   POOL KIND
org-giantswarm   pssss   wc-cp-ips   GlobalInClusterIPPool

NAMESPACE        NAME    ADDRESS         POOL NAME   POOL KIND

> org-giantswarm   pssss   10.10.222.233   wc-cp-ips   GlobalInClusterIPPool

Reviewer: Is there a command line missing here? One `kubectl get` wouldn't produce 2 header rows in this case.

Author: yep

```

> You can find more details about how it works in [this blog](https://deploy-preview-8--jkremser.netlify.app/post/ipam-for-capv/).

Reviewer 2: ```suggestion
You can find more details about how it works in this [blog post](https://kremser.dev/post/ipam-for-capv/).
```

> ## Slicing the subnet

Reviewer: Just an idea: would "dividing" be a better-understood term compared to "slicing"?

Here we will talk about how to "slice" the subnet where you will be deploying your workload clusters (provided they will all be on the same subnet of course).

The main goal here is to avoid IP address collisions so we want the subnet to accomodate for:

* DHCP range for node interface.
* One IP for the management clusters' Kubernetes API.

> * One IP for the management clusters' Ingress.

Reviewer: "load balancer IP" maybe? (if I understand correctly that this is the floating IP used by kube-vip)

Author: That would be a bit confusing since it is an IP on the subnet.
Maybe 
`One IP for the management clusters' xxx for use by kube-vip`

* Pool of IPs for workload clusters' Kubernetes API.
* Pool of IPs for workload clusters' Ingresses (untracked by IPAM controller).

Below is an example of what it could look like but you are free to adjust the ranges as you see fit:

> ![kube-vip range](capv-kubevip-ipam.png)

Reviewer: We normally provide the editable image source so others can edit the diagram later.

For example:

```
<!-- Source for above image: https://docs.google.com/drawings/d/1PDve3HoE7br_6npe0RSCw8ddt-H7pFztPlSDiQbNabs/edit -->
```

Or alternatively, if the diagram is from draw.io, export a PNG including the source.

Author: It's from excalidraw (like draw.io).
TIL you can embed the source in the png

## IPs of Giant Swarm workload clusters

### Control plane

When deploying a CAPV workload cluster, you don't need to select an IP address for the control plane, you can leave the `host` field empty and specify a `globalinclusterippools` instead. If there are free IPs in the pool, the pre-install job will mutate the cluster with the claimed IP address.

```yaml
connectivity:
  network:
    controlPlaneEndpoint:
      host: ""
      ipPoolName: wc-cp-ips
      port: 6443
```

### Services of type Load Balancer

> Giant Swarm clusters are deployed with the `Kube-vip` cloud provider interface (CPI), meaning you can use an IP address from the underlying layer 2 subnet for services of type Load Balancer. However, as mentioned previously, those are not tracked by the `cluster-api-ipam-provider-in-cluster` controller, hence the importance of properly slicing your subnet.

Reviewer: ```suggestion
Giant Swarm clusters are deployed with the `Kube-vip` cloud provider interface (CPI), meaning you can use an IP address from the underlying layer 3 subnet for `Service` objects of `type: LoadBalancer`. However, as mentioned previously, those are not tracked by the `cluster-api-ipam-provider-in-cluster` controller, hence the importance of properly slicing your subnet.
```

?

---
linkTitle: vSphere Kube-vip
title: vSphere networking with Kube-vip
description: Here we describe how to work with Kube-vip in Giant Swarm CAPV clusters.
weight: 25
menu:
  main:
    parent: advanced
last_review_date: 2023-10-09
user_questions:
  - How to divide my subnet for Kube-vip with CAPV?
  - What IP should I use for Kube-vip with CAPV?
  - How to set a range of IPs for load balancer in Kube-vip with CAPV?
  - Where can I see the IPs in use by Kube-vip?
aliases:
  - /guides/kube-vip-for-capv/
owner:
  - https://github.com/orgs/giantswarm/teams/team-rocket
---

> While most infrastructure providers offer a load balancer to use for the Kubernetes API and services of type Load Balancer (AWS, Azure, VMware Cloud Director...), vSphere is a bit different in that it is a hypervisor as opposed to a cloud provider. Advanced networking use cases such as routing, firewalling, load balancing and the likes are handled by VMware NSX. In order to offer a highly-available Kubernetes API even to customers without NSX, CAPV comes with `kube-vip`, a layer-2 load balancer.

Reviewer 3: ```suggestion

While most infrastructure providers offer a load balancer to use for the Kubernetes API and services of type Load Balancer (AWS, Azure, VMware Cloud Director...), vSphere is a bit different in that it is a hypervisor as opposed to a cloud provider. Advanced networking use cases such as routing, firewalling, load balancing and the likes are handled by VMware NSX. In order to offer a highly-available Kubernetes API even to customers without NSX, CAPV comes with `kube-vip`, a layer-2 load balancer.
```

As a result of working on layer 2, `Kube-vip` uses IP addresses in the same subnet as the nodes themselves and these IP addresses must be selected carefully to avoid collisions. This is particularly plausible if multiple users collaborate on the platform and select the same IP for different clusters.

## IPAM controller

> in order to mitigate this risk, Giant Swarm CAPV clusters run `cluster-api-ipam-provider-in-cluster` to track IP addresses usage from within the management cluster (MC). For that reason, **customers shouldn't run more than one management cluster per subnet**.

Reviewer 3: ```suggestion
In order to mitigate this risk, Giant Swarm CAPV clusters run `cluster-api-ipam-provider-in-cluster` to track IP address usage from within the management cluster (MC). For that reason, **customers shouldn't run more than one management cluster per subnet**.
```

Reviewer 3: But this is something customer run or we do for them?

Author: This is us, we added it to the cluster chart

NAMESPACE        NAME    POOL NAME   POOL KIND
org-giantswarm   pssss   wc-cp-ips   GlobalInClusterIPPool

$ kubectl get ipaddresses.ipam.cluster.x-k8s.io -A
NAMESPACE        NAME    ADDRESS         POOL NAME   POOL KIND
org-giantswarm   pssss   10.10.222.233   wc-cp-ips   GlobalInClusterIPPool
```

You can find more details about how it works in this [blog post](https://kremser.dev/post/ipam-for-capv/).

## Dividing the subnet

Here we will talk about how to "divide" the subnet where you will be deploying your workload clusters (provided they will all be on the same subnet of course).

The main goal here is to avoid IP address collisions so we want the subnet to accomodate for:

* DHCP range for node interface.
* One IP for the management clusters' Kubernetes API for use by kube-vip.
* One IP for the management clusters' Ingress for use by kube-vip.
* Pool of IPs for workload clusters' Kubernetes API.
* Pool of IPs for workload clusters' Ingresses (untracked by IPAM controller).

Below is an example of what it could look like but you are free to adjust the ranges as you see fit:

![kube-vip range](capv-kubevip-ipam.png)

## Workload cluster IPs

### Control plane

> When deploying a CAPV workload cluster, you don't need to select an IP address for the control plane, you can leave the `host` field empty and specify a `globalinclusterippools` instead. If there are free IPs in the pool, the pre-install job will mutate the cluster with the claimed IP address.

Reviewer 3: ```suggestion
When creating a CAPV workload cluster, you don't need to select an IP address for the control plane, you can leave the `host` field empty and specify a `globalinclusterippools` instead. If there are free IPs in the pool, the pre-install job will mutate the cluster with the claimed IP address.
```
I like more create in this case, WDYT?

Author: Sound good to me
